### PR TITLE
install fails if enroll fails and surface errors

### PR DESCRIPTION
**Pull request by AndersonQ**
For the agent to be actually enrolled it needs to restart after the enroll process is completed, so it'll pickup the new config and "connect" to fleet-server.

This change makes the enroll command to fail if it cannot restart the agent after enrolling on fleet

## What does this PR do?

This change makes the enroll command to fail if it cannot restart the agent after enrolling on fleet

## Why is it important?

For the agent to be actually enrolled it needs to restart after the enroll process is completed, so it'll pickup the new config and "connect" to fleet-server.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Related issues

- N/A

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: changelog/fragments/1693403216-Surface-errors-during-Agent's-enroll.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# - breaking-change: a change to previously-documented behavior
+# - deprecation: functionality that is being removed in a later release
+# - bug-fix: fixes a problem in a previous version
+# - enhancement: extends functionality but does not break or fix existing behavior
+# - feature: new functionality
+# - known-issue: problems that we are aware of in a given version
+# - security: impacts on the security of a product or a user’s deployment.
+# - upgrade: important information for someone upgrading from a prior version
+# - other: does not fit into any of the other categories
+kind: feature
```
Comment (ycombinator) [inline changelog/fragments/1693403216-Surface-errors-during-Agent's-enroll.yaml:11]: I think this should probably be a `bug-fix`? Seems like it's a bug for `elastic-agent install --url=... --enrollment-token-...` to succeed if enrollment fails. 
```suggestion
kind: bug-fix
```

File: internal/pkg/agent/cmd/enroll.go [hunk 2]
```diff
@@ -351,7 +351,7 @@ func enroll(streams *cli.IOStreams, cmd *cobra.Command) error {
 	//  Error: failed to fix permissions: chown /Library/Elastic/Agent/data/elastic-agent-c13f91/elastic-agent.app: operation not permitted
 	// This is because we are fixing permissions twice, once during installation and again during the enrollment step.
 	// When we are enrolling as part of installation on MacOS, skip the second attempt to fix permissions.
-	var fixPermissions bool = fromInstall
+	var fixPermissions = fromInstall
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/enroll.go:354]: Might as well go all the way here?
```suggestion
	fixPermissions := fromInstall
```

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 3]
```diff
@@ -267,17 +268,23 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
...
 	}()
 
 	if c.agentProc == nil {
-		if err := c.daemonReload(ctx); err != nil {
-			c.log.Infow("Elastic Agent might not be running; unable to trigger restart", "error", err)
-		} else {
-			c.log.Info("Successfully triggered restart on running Elastic Agent.")
+		if err = c.daemonReloadWithBackoff(ctx); err != nil {
+			c.log.Errorf("Elastic Agent might not be running; unable to trigger restart: %v", err)
+			return fmt.Errorf("could not reload agent deamon, unable to trigger restart: %w", err)
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/enroll_cmd.go:281]: Given the error printing that's now happening in the `defer func()` above, would it make sense to set `err` to the wrapped error here and return that instead?
```suggestion
			err = fmt.Errorf("could not reload agent deamon, unable to trigger restart: %w", err)
			return err
```

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 4]
```diff
@@ -267,17 +268,23 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
 	}
 
 	defer func() {
-		fmt.Fprintln(streams.Out, "Successfully enrolled the Elastic Agent.")
+		if err != nil {
+			fmt.Fprintf(streams.Out, "Something went wrong while enrolling the Elastic Agent: %v\n", err)
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/enroll_cmd.go:272]: Should this error be printed to `streams.Err`?
```suggestion
			fmt.Fprintf(streams.Err, "Something went wrong while enrolling the Elastic Agent: %v\n", err)
```

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 5]
```diff
@@ -478,8 +485,20 @@ func (c *enrollCmd) enrollWithBackoff(ctx context.Context, persistentConfig map[
...
 	c.log.Infof("Starting enrollment to URL: %s", c.client.URI())
 	err := c.enroll(ctx, persistentConfig)
+	if err == nil {
+		return nil
+	}
+
+	const deadline = 10 * time.Minute
+	const frequency = 60 * time.Second
+
+	c.log.Infof("1st enrollment attempt failed, retrying for %s each %s enrolling to URL: %s",
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/enroll_cmd.go:495]:
```suggestion
	c.log.Infof("1st enrollment attempt failed, retrying for %s, every %s, enrolling to URL: %s",
```

File: internal/pkg/agent/cmd/install.go [hunk 6]
```diff
@@ -83,7 +83,7 @@ func installCmd(streams *cli.IOStreams, cmd *cobra.Command) error {
 
 	nonInteractive, _ := cmd.Flags().GetBool("non-interactive")
 	if nonInteractive {
-		fmt.Fprintf(streams.Out, "Installing in non-interactive mode.")
+		fmt.Fprintf(streams.Out, "Installing in non-interactive mode.\n")
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/install.go:86]:
```suggestion
		fmt.Fprintln(streams.Out, "Installing in non-interactive mode.")
```

File: internal/pkg/agent/cmd/install.go [hunk 7]
```diff
@@ -205,6 +205,8 @@ func installCmd(streams *cli.IOStreams, cmd *cobra.Command) error {
 				}
 			}()
 		}
+
+		fmt.Fprint(streams.Out, "Elastic Agent successfully installed, starting enrollment.\n")
```
Comment (ycombinator) [inline internal/pkg/agent/cmd/install.go:209]:
```suggestion
		fmt.Fprintln(streams.Out, "Elastic Agent successfully installed, starting enrollment.")
```

**Review by ycombinator — COMMENTED** [review]
Nice improvement! Left a few minor comments.

File: internal/pkg/agent/cmd/enroll_cmd.go (hunk 4 above)
Comment (AndersonQ) [reply internal/pkg/agent/cmd/enroll_cmd.go]: I did not see `streams.Err` being used here, so I just followed the pattern. But I agree with you

File: internal/pkg/agent/cmd/enroll_cmd.go (hunk 3 above)
Comment (AndersonQ) [reply internal/pkg/agent/cmd/enroll_cmd.go:281]: they have the same effect: https://go.dev/play/p/9JnbEQWsxCG, err "will be set by returning the error"

**Review by ycombinator — APPROVED** [review]

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 8]
```diff
@@ -478,8 +493,20 @@ func (c *enrollCmd) enrollWithBackoff(ctx context.Context, persistentConfig map[
 
 	c.log.Infof("Starting enrollment to URL: %s", c.client.URI())
 	err := c.enroll(ctx, persistentConfig)
+	if err == nil {
+		return nil
+	}
+
+	const deadline = 10 * time.Minute
+	const frequency = 60 * time.Second
```
Comment (cmacknz) [inline internal/pkg/agent/cmd/enroll_cmd.go:501]: This feels too slow for me. We can retry faster than once every minute.

I also think waiting for 10 minutes is too long, we should fail earlier than this.

Right now the enroll command is going to appear like it is hung since there won't be any user feedback faster than every 1 minute.

Comment (cmacknz) [reply internal/pkg/agent/cmd/enroll_cmd.go:501]: Ah I see this is what it was before for the Fleet URL, I care more about the new backoff  for the restart command on the same host. Commented in the wrong place.

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 9]
```diff
@@ -443,24 +450,32 @@ func (c *enrollCmd) prepareFleetTLS() error {
 
 func (c *enrollCmd) daemonReloadWithBackoff(ctx context.Context) error {
 	err := c.daemonReload(ctx)
+	if err != nil &&
```
Comment (cmacknz) [inline internal/pkg/agent/cmd/enroll_cmd.go:453]: Why do we need to special case this? Why can't we just enter the backoff loop below and exit it once it is finished?

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 10]
```diff
@@ -443,24 +450,32 @@ func (c *enrollCmd) prepareFleetTLS() error {
...
-	for i := 5; i >= 0; i-- {
+	var i int
+	for ; i < 5; i++ {
 		backExp.Wait()
 		c.log.Info("Retrying to restart...")
 		err = c.daemonReload(ctx)
-		if err == nil {
+		if err == nil ||
+			errors.Is(err, context.DeadlineExceeded) ||
+			errors.Is(err, context.Canceled) {
```
Comment (cmacknz) [inline internal/pkg/agent/cmd/enroll_cmd.go:472]: You could replace this with https://pkg.go.dev/github.com/cenkalti/backoff/v4 which has a `WithContext` option to natively understands contexts. We already use this package for upgrades. Example:

https://github.com/cmacknz/elastic-agent/blob/261b94e5ec8372dda376ac80cdd212902b3a5aba/internal/pkg/agent/application/upgrade/step_download.go#L155-L185

File: internal/pkg/agent/cmd/enroll_cmd.go [hunk 11]
```diff
@@ -267,17 +268,23 @@ func (c *enrollCmd) Execute(ctx context.Context, streams *cli.IOStreams) error {
...
+			fmt.Fprintln(streams.Out, "Successfully enrolled the Elastic Agent.")
+		}
 	}()
 
 	if c.agentProc == nil {
-		if err := c.daemonReload(ctx); err != nil {
-			c.log.Infow("Elastic Agent might not be running; unable to trigger restart", "error", err)
-		} else {
-			c.log.Info("Successfully triggered restart on running Elastic Agent.")
+		if err = c.daemonReloadWithBackoff(ctx); err != nil {
```
Comment (cmacknz) [inline internal/pkg/agent/cmd/enroll_cmd.go:279]: I think the 1 minute retry with a 10 minute wait might be too slow for attempts to restart the agent on the same host. We can likely retry every second to start with a maximum wait of up to 2 minutes. If we are still retrying after 2 minutes changes are we aren't going to get the agent to restart.

We seem to need this logic in integration tests to get them to pass reliably and using a 1 minute minimum interval is going to make every test that hits this case take an additional minute to run.

The same is probably true for the Fleet retries but that one is tied to system scalability during mass enrollments so I am more hesitant to change it.

**Comment by AndersonQ** [comment]
buildkite test this

**Comment by cmacknz** [comment]
You will need to merge in the latest changes in main to get the tests to run properly if you haven't done that already. We had to change the region we provisioned the stack in.

**Comment by cmacknz** [comment]
This is failing with the error it could be fixing:

```
    endpoint_security_test.go:189: Installing in non-interactive mode.
        Copying files................................................... DONE
        Installing service.... DONE
        Starting service... DONE
        Elastic Agent successfully installed, starting enrollment.
        Enrolling Elastic Agent with Fleet....{"log.level":"info","@timestamp":"2023-10-05T17:49:49.581Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":494},"message":"Starting enrollment to URL: https://2bb73e932bae48fba9bd4ac7037849c5.fleet.us-central1.gcp.foundit.no:443/","ecs.version":"1.6.0"}
        .............................................................................{"log.level":"info","@timestamp":"2023-10-05T17:50:10.808Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":468},"message":"Retrying to restart...","ecs.version":"1.6.0"}
        {"log.level":"error","@timestamp":"2023-10-05T17:50:10.826Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":280},"message":"Elastic Agent might not be running; unable to trigger restart: could not reload deamon after 1 retries: %!w()","ecs.version":"1.6.0"}
        Something went wrong while enrolling the Elastic Agent: could not reload deamon after 1 retries: %!w()
        Error: could not reload agent daemon, unable to trigger restart: could not reload deamon after 1 retries: %!w()
        For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.10/fleet-troubleshooting.html
         FAILED
        Stopping service... DONE
        Uninstalling...
           Stopping service... DONE
           Stopping upgrade watcher; none found... DONE
        .   Removing service.... DONE
           Removing install directory... DONE
           DONE
        Error: enroll command failed for unknown reason: exit status 1
        For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.10/fleet-troubleshooting.html
        
    endpoint_security_test.go:191: 
        	Error Trace:	/home/ubuntu/agent/testing/integration/endpoint_security_test.go:191
        	            				/home/ubuntu/agent/testing/integration/endpoint_security_test.go:88
        	Error:      	Received unexpected error:
        	            	unable to enroll Elastic Agent: error running agent install command: exit status 1
        	Test:       	TestInstallAndCLIUninstallWithEndpointSecurity/protected
        	Messages:   	failed to install agent with policy
--- FAIL: TestInstallAndCLIUninstallWithEndpointSecurity/protected (55.99s)
```

We need more than 1 retry when we can't restart the agent to fix this.

**Comment by cmacknz** [comment]
> Error: enroll command failed for unknown reason: exit status 1

Also that error message can be improved since we know exactly why enroll failed.
